Diagnostics for opened files

tomv564 opened the pull request:
This is a small implementation towards #162:
* ProjectManager exposes files to compiler via ProjectConfiguration in didOpen/didChange
* ProjectManager passes a DiagnosticsHandler to ProjectConfiguration
* ~~Expose `programSyncEvents` on ProjectConfiguration, triggered after compilation~~
* ~~DiagnosticsPublisher requests diagnostics on a synced program and publishes per file~~
* ProjectConfiguration passes TS file diagnostics to DiagnosticsHandler 
* ~~Basic tests~~
  * Unit tests for DiagnosticsPublisher
  * Existing functional tests expanded to test correct uri and content update handling

> Uncertainties:
> * It feels totally reasonable to expose a file to the compiler on didOpen, but for some reason this was not being done yet?
> * As discussed, VS Code stops showing diagnostics on broken code when you close the file. We would have to remove the file from the compiler to achieve consistent behaviour with VS Code.
> * Missing test scenarios for broken tsconfig, missing node modules, etc.
> * A minimum delay of 1 second feels good on instantly-compiling projects, but perhaps it should be the editor's job to delay showing feedback while the user is typing? Or should the didChange handler not trigger compiles immediately?

Thanks for the answers on these! Continuing conversation below.

felixfbecker commented:
It is quite hard for me to understand what is going on here (some documentaiton on DiagnosticsPublisher would be helpful). What I got:

- ProjectConfiguration now emits an event with a new project whenever it is updated
- The ProjectManager subscribes to this event, but it does so by being passed to ProjectManager, which passes it to ProjectConfiguration, which then passes an event Subject to a method.
- The DiagnosticsPublisher takes a Subscribable to subscribe on and maintains a map of file to collected diagnostics. On a program update, it updates the map, then publishes them all.

Problems I have:
- The point of evented API is that the event producer is not coupled to (doesn't know about) the event consumer. The producer just emits the event, the consumer subscribes to it.
- The `subscribe` method looks like the the one defined in the Rx `Subscribable` interface, but does the exact opposite. It doesn't take a Subscriber, it takes a Subscribable to subscribe on.
- Subjects are an anti pattern. If you cannot avoid them, their scope should be limited - publicly exposed should be only the Observable interface.
- I think using an Observable/Subject for the program update event is overkill. Are there any benefits over making the ProjectConfiguration and ProjectManager a simple EventEmitter?
- Tests: The unit test for ProjectManager should just test the unit under test (ProjectManager). That means it should just assert that the program update event is emitted.

> It feels totally reasonable to expose a file to the compiler on didOpen, but for some reason this was not being done yet?

No reason, makes sense to me. I think it didn't matter because we read all files in anyway and all operations ensured the files they needed.

> As discussed, VS Code stops showing diagnostics on broken code when you close the file. We would have to remove the file from the compiler to achieve consistent behaviour with VS Code.

I wouldn't be too worried about that, I don't have a problem with showing diagnostics for closed files.

> A minimum delay of 1 second feels good on instantly-compiling projects, but perhaps it should be the editor's job to delay showing feedback while the user is typing? Or should the didChange handler not trigger compiles immediately?

I would expect the client to not spam `didChange` on every key stroke, so I don't see that much of an issue with publishing diagnostics immediately.

felixfbecker commented:
Please ping when this is ready for another review

tomv564 commented:
I've addressed points from the initial review, thanks for taking the time!

> It is quite hard for me to understand what is going on here
* Added docstrings, comments and broke up DiagnosticsPublisher's update function

> Using Rx is overkill
* Agreed, ProjectConfiguration now just calls `updateFileDiagnostics` on DiagnosticsPublisher

> The unit test for ProjectManager should just test the unit under test (ProjectManager)
* Created unit tests to test the details of DiagnosticsPublisher's work
* Kept the functional tests (titled 'Diagnostics') to catch regressions in the system

I have some time to work on this, but feel free to work on my branch!

tomv564 commented:
Ping @felixfbecker, ready for another review!

felixfbecker commented on src/project-manager.ts:
```diff
@@ -951,6 +971,28 @@ export class ProjectConfiguration {
+	 * Ensures a single file is available to the LanguageServiceHost
+	 * @param filePath
+	 */
+	ensureSourceFile(filePath: string): boolean {
```
As I understand it, the return type indicates whether the program was resynced as part of the method call? Why exactly is that information needed for a caller who wants to ensure a source file? If it is needed to decide whether to republish diagnostics, then this puts unnecessary burden on the callers to always check the return value and act on it - emitting an event is better in that case

felixfbecker commented on src/test/diagnostics.test.ts:
```diff
@@ -0,0 +1,219 @@
+
+	beforeEach(() => {
+		publishSpy = sinon.spy();
+		langClient = {
```
Since you want to stub all and only spy on one I would use `sinon.createStubInstance(RemoteLanguageClient)`

tomv564 replied:
Awesome!

felixfbecker commented on src/test/diagnostics.test.ts:
```diff
@@ -0,0 +1,219 @@
+});
+
+// Integration tests that include TS compilation.
+describe('Diagnostics', () => {
```
Still not convinced we should have these here.

The expected behaviour of the ProjectManager is to call the DiagnosticHandler, so there should be tests in the ProjectManager suite that assert it is called. That's where all the setup/teardown code is located.

All tests are to prevent regressions

tomv564 replied:
Right you are, these have been replaced with some unit tests on didChange/didOpen.

tomv564 commented:
@felixfbecker thanks for the guidance, hope the didChange and test changes are what you intended!

felixfbecker commented on src/project-manager.ts:
```diff
@@ -464,9 +471,10 @@ export class ProjectManager implements Disposable {
+		config.ensureSourceFile(filePath);
 		config.getHost().incProjectVersion();
-		config.syncProgram(span);
+		config.syncProgram();
```
why the removal of the span?

felixfbecker commented on src/diagnostics.ts:
```diff
@@ -0,0 +1,118 @@
+		this.problemFiles.clear();
+
+		// for each file: publish and set as problem file
+		diagnosticsByFile.forEach((diagnostics, file) => {
```
`for of`

felixfbecker commented on src/test/diagnostics.test.ts:
```diff
@@ -0,0 +1,84 @@
+
+describe('DiagnosticsPublisher', () => {
+	let langClient: LanguageClient;
+	let diagnosticsManager;
```
no type

felixfbecker commented on src/test/diagnostics.test.ts:
```diff
@@ -0,0 +1,84 @@
+	let langClient: LanguageClient;
+	let diagnosticsManager;
+	const createTSFileDiagnostic = (message: string, file: ts.SourceFile) => {
+		return {
```
use arrow function without return or named function

felixfbecker commented on src/test/diagnostics.test.ts:
```diff
@@ -0,0 +1,84 @@
+	beforeEach(() => {
+		publishSpy = sinon.spy();
+		langClient = sinon.createStubInstance(RemoteLanguageClient);
+		langClient.textDocumentPublishDiagnostics = publishSpy;
```
all methods are stubs now, so no need to assign the spy here. Just type the langClient as something like
```ts
{ [K in keyof LanguageClient]: LanguageClient[K] & sinon.SinonStub }
// or
LanguageClient & { textDocumentPublishDiagnostics: Function & sinon.SinonSpy }

felixfbecker commented on src/test/project-manager.test.ts:
```diff
@@ -70,4 +75,46 @@ describe('ProjectManager', () => {
+		it('should compile opened file and return diagnostics', async () => {
+			projectManager.didOpen('file:///src/dummy.ts', 'const num: number = "banana";');
+			const lastArgs = diagnosticsSpy.lastCall.args[0];
+			assert.equal(lastArgs.length, 1);
```
Pretty sure there is a sinon assert function for this. Should also first assert that the function is called at all?

tomv564 replied:
Added an assert as suggested, but I couldn't find a way to assert on the last call only with sinon's asserts. They suggest using `sinon.assert.calledWith`, and we would need a `sinon.match` to avoid check all of TS's  diagnostics type. I tried this briefly but ran out of time, let me know if you want me to switch it to this syntax.

felixfbecker commented on src/typescript-service.ts:
```diff
@@ -67,7 +68,6 @@ export type TypeScriptServiceFactory = (client: LanguageClient, options?: TypeSc
  * underscore.
  */
 export class TypeScriptService {
-
```
revert

felixfbecker commented on src/diagnostics.ts:
```diff
@@ -0,0 +1,118 @@
+	 * Requires a connection to the remote client to send diagnostics to
+	 * @param remoteClient
+	 */
+	constructor(private remoteClient: LanguageClient) {}
```
rename to `client`

felixfbecker commented on src/diagnostics.ts:
```diff
@@ -0,0 +1,118 @@
+	 * The files that were last reported to have errors
+	 * If they don't appear in the next update, we must publish empty diagnostics for them.
+	 */
+	private problemFiles: Set<string> = new Set();
```
prefer type inference

felixfbecker commented on src/diagnostics.ts:
```diff
@@ -0,0 +1,118 @@
+		const diagnosticsByFile = new Map<string, ts.Diagnostic[]>();
+		for (const diagnostic of diagnostics) {
+			// TODO for some reason non-file diagnostics end up in here (where file is undefined)
+			const fileName = diagnostic.file ? diagnostic.file.fileName : '';
```
`const fileName = diagnostic.file && diagnostic.file.fileName`

tomv564 commented:
Thanks for the review, apologies for the unintended changes!

felixfbecker commented:
Thanks! Great work
